### PR TITLE
Record: CROWN-Q + GPTQ + Legal TTT — val_bpb 1.1174 (3-seed mean)

### PR DESCRIPTION
**Pull request by EthanYangTW**
## Summary

11L GQA + XSA-all + full Cholesky GPTQ + score-first AdamW TTT. Sqrt cooldown schedule holds LR higher during warmdown, improving post-quantization TTT quality.

**val_bpb: 1.1174 (3-seed mean, std 0.0004)**

## Results

| Seed | TTT BPB | Artifact |
|------|---------|----------|
| 1337 | **1.1170** | 15,961,751 bytes |
| 42 | **1.1176** | 15,850,151 bytes |
| 7 | **1.1176** | 15,844,080 bytes |
| **Mean** | **1.1174** | |
| **Std** | **0.0004** | |

## Key Techniques

- **Full Cholesky GPTQ**: Hessian-aware quantization with act-order, self-generated calibration (no val data).
- **Score-first AdamW TTT**: Each token scored before any gradient update using it. Last 2 blocks unfrozen (4.7M/27M params), 3 epochs.
- **Sqrt cooldown**: `sqrt(x)` schedule during warmdown instead of linear. Holds LR higher longer.
- **CROWN-Q**: Curvature-weighted quantization variance penalty during warmdown.

## Architecture

- 11L, 512d, GQA 8/4, MLP 3x relu²
- XSA on all 11 layers, BigramHash 2048
- Partial RoPE 16/64, SmearGate + OrthoInit
- EMA 0.997, SWA, Late QAT at 50% warmdown
- 26.99M params, int6 + zstd-22

## Timing

- Training: 600s wallclock (8xH100 SXM), 89ms/step (FA3 Hopper), ~6650 steps
- Eval: sliding window stride=32 (~150s) + TTT 3 epochs (~460s) ≈ 610s

## Compliance

- Training ≤ 600s wallclock
- GPTQ calibration: self-generated tokens (no validation data used)
- TTT: legal score-first (every token scored before any gradient update)
- All artifacts < 16,000,000 bytes

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: CROWN-Q + GPTQ + Legal TTT — val_bpb 1.1174 (3-seed mean)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

## Summary

PR #1129 ("V38_SqrtCooldown_3seed") submits a post-quantization sliding-window TTT evaluation function `eval_val_sliding_ttt` (lines 894–1064).

## N-gram / BigramHash check

`BigramHashEmbedding.bigram_hash` (lines 632–637) hashes consecutive *input* tokens:
```
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```
`t` is derived from `token_ids` (the input sequence), **not** from targets. No target token is XOR'd into the hash key. The ILLEGAL n-gram family bug (target-in-hash) is **absent**.

## TTT legality check

`eval_val_sliding_ttt` operates in two clearly separated phases per chunk:

**Phase 1 (lines 970–1005):** Score all windows in the current chunk under `torch.no_grad()` / `base_model.eval()`. NLL for scored tokens is accumulated into `loss_sum`. No gradient flows here.

**Phase 2 (lines 1007–1045):** Train on the *already-scored* chunk. The `is_last_chunk` guard (line 1008) ensures the final chunk is **never** trained on — no future-data contamination. This is exactly the PR #1413 pattern.

The function is called post-quantization only (line 1751: `# === EVAL PHASE: Post-quant TTT...`). The explicit comment at line 1687 reads: `# Export best weights (no pre-quant TTT — that's illegal per PR #606 precedent)`, confirming pre-quant TTT is intentionally absent.

## Scored-region SLOT

No scored-region SLOT manipulation found. The sliding window scores only the stride-sized new tokens per window (line 999: `nll[i, s:wlen]`), which is standard.

## Conclusion

All three legal requirements are satisfied: (1) no target-XOR hash key, (2) score-first TTT with `is_last_chunk` guard, (3) no pre-quant TTT. The submission is clean.

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
